File: src/metrics.rs
use serde_json::{Map, Value};
use std::ops::AddAssign;

#[derive(Debug, Default, Copy, Clone)]
pub struct Metrics {
    n: f64,
    nargs: f64,
    nexits: f64,
    cognitive: f64,
    cyclomatic: f64,
    halstead_n1: f64,
    halstead_capital_n1: f64,
    halstead_n2: f64,
    halstead_capital_n2: f64,
    halstead_length: f64,
    halstead_estimated_program_length: f64,
    halstead_purity_ratio: f64,
    halstead_vocabulary: f64,
    halstead_volume: f64,
    halstead_difficulty: f64,
    halstead_level: f64,
    halstead_effort: f64,
    halstead_time: f64,
    halstead_bugs: f64,
    sloc: f64,
    ploc: f64,
    lloc: f64,
    cloc: f64,
    blank: f64,
    functions: f64,
    closures: f64,
    mi_original: f64,
    mi_sei: f64,
    mi_visual_studio: f64,
}

impl Metrics {
    fn get_metric(metrics: &Map<String, Value>, key1: &str, key2: &str) -> f64 {
        metrics[key1].as_object().unwrap()[key2]
            .as_f64()
            .unwrap_or_default()
    }

    pub fn from_value(value: &Value) -> Self {
        let metrics = value.as_object().unwrap();
        Metrics {
            n: 0.0,
            nargs: Self::get_metric(metrics, "nargs", "sum"),
            nexits: Self::get_metric(metrics, "nexits", "sum"),
            cognitive: Self::get_metric(metrics, "cognitive", "sum"),
            cyclomatic: Self::get_metric(metrics, "cyclomatic", "sum"),
            halstead_n1: Self::get_metric(metrics, "halstead", "n1"),
            halstead_capital_n1: Self::get_metric(metrics, "halstead", "N1"),
            halstead_n2: Self::get_metric(metrics, "halstead", "n2"),
            halstead_capital_n2: Self::get_metric(metrics, "halstead", "N2"),
            halstead_length: Self::get_metric(metrics, "halstead", "length"),
            halstead_estimated_program_length: Self::get_metric(
                metrics,
                "halstead",
                "estimated_program_length",
            ),
            halstead_purity_ratio: Self::get_metric(metrics, "halstead", "purity_ratio"),
            halstead_vocabulary: Self::get_metric(metrics, "halstead", "vocabulary"),
            halstead_volume: Self::get_metric(metrics, "halstead", "volume"),
            halstead_difficulty: Self::get_metric(metrics, "halstead", "difficulty"),
            halstead_level: Self::get_metric(metrics, "halstead", "level"),
            halstead_effort: Self::get_metric(metrics, "halstead", "effort"),
            halstead_time: Self::get_metric(metrics, "halstead", "time"),
            halstead_bugs: Self::get_metric(metrics, "halstead", "bugs"),
            sloc: Self::get_metric(metrics, "loc", "sloc"),
            ploc: Self::get_metric(metrics, "loc", "ploc"),
            lloc: Self::get_metric(metrics, "loc", "lloc"),
            cloc: Self::get_metric(metrics, "loc", "cloc"),
            blank: Self::get_metric(metrics, "loc", "blank"),
            functions: Self::get_metric(metrics, "nom", "functions"),
            closures: Self::get_metric(metrics, "nom", "closures"),
            mi_original: Self::get_metric(metrics, "mi", "mi_original"),
            mi_sei: Self::get_metric(metrics, "mi", "mi_sei"),
            mi_visual_studio: Self::get_metric(metrics, "mi", "mi_visual_studio"),
        }
    }

    pub fn avg(&self) -> Self {
        Metrics {
            n: 0.,
            nargs: self.nargs / self.n,
            nexits: self.nexits / self.n,
            cognitive: self.cognitive / self.n,
            cyclomatic: self.cyclomatic / self.n,
            halstead_n1: self.halstead_n1 / self.n,
            halstead_capital_n1: self.halstead_capital_n1 / self.n,
            halstead_n2: self.halstead_n2 / self.n,
            halstead_capital_n2: self.halstead_capital_n2 / self.n,
            halstead_length: self.halstead_length / self.n,
            halstead_estimated_program_length: self.halstead_estimated_program_length / self.n,
            halstead_purity_ratio: self.halstead_purity_ratio / self.n,
            halstead_vocabulary: self.halstead_vocabulary / self.n,
            halstead_volume: self.halstead_volume / self.n,
            halstead_difficulty: self.halstead_difficulty / self.n,
            halstead_level: self.halstead_level / self.n,
            halstead_effort: self.halstead_effort / self.n,
            halstead_time: self.halstead_time / self.n,
            halstead_bugs: self.halstead_bugs / self.n,
            sloc: self.sloc / self.n,
            ploc: self.ploc / self.n,
            lloc: self.lloc / self.n,
            cloc: self.cloc / self.n,
            blank: self.blank / self.n,
            functions: self.functions / self.n,
            closures: self.closures / self.n,
            mi_original: self.mi_original / self.n,
            mi_sei: self.mi_sei / self.n,
            mi_visual_studio: self.mi_visual_studio / self.n,
        }
    }

    pub fn weigh(&self, weight: f64) -> Self {
        Metrics {
            n: self.n,
            nargs: self.nargs,
            nexits: self.nexits,
            cognitive: self.cognitive * weight,
            cyclomatic: self.cyclomatic * weight,
            halstead_n1: self.halstead_n1 * weight,
            halstead_capital_n1: self.halstead_capital_n1 * weight,
            halstead_n2: self.halstead_n2 * weight,
            halstead_capital_n2: self.halstead_capital_n2 * weight,
            halstead_length: self.halstead_length * weight,
            halstead_estimated_program_length: self.halstead_estimated_program_length * weight,
            halstead_purity_ratio: self.halstead_purity_ratio * weight,
            halstead_vocabulary: self.halstead_vocabulary * weight,
            halstead_volume: self.halstead_volume * weight,
            halstead_difficulty: self.halstead_difficulty * weight,
            halstead_level: self.halstead_level * weight,
            halstead_effort: self.halstead_effort * weight,
            halstead_time: self.halstead_time * weight,
            halstead_bugs: self.halstead_bugs * weight,
            sloc: self.sloc,
            ploc: self.ploc,
            lloc: self.lloc,
            cloc: self.cloc,
            blank: self.blank,
            functions: self.functions,
            closures: self.closures,
            mi_original: self.mi_original * weight,
            mi_sei: self.mi_sei * weight,
            mi_visual_studio: self.mi_visual_studio * weight,
        }
    }
}

impl AddAssign for Metrics {
    fn add_assign(&mut self, other: Self) {
        self.n += 1.0;
        self.nargs += other.nargs;
        self.nexits += other.nexits;
        self.cognitive += other.cognitive;
        self.cyclomatic += other.cyclomatic;
        self.halstead_n1 += other.halstead_n1;
        self.halstead_capital_n1 += other.halstead_capital_n1;
        self.halstead_n2 += other.halstead_n2;
        self.halstead_capital_n2 += other.halstead_capital_n2;
        self.halstead_length += other.halstead_length;
        self.halstead_estimated_program_length += other.halstead_estimated_program_length;
        self.halstead_purity_ratio += other.halstead_purity_ratio;
        self.halstead_vocabulary += other.halstead_vocabulary;
        self.halstead_volume += other.halstead_volume;
        self.halstead_difficulty += other.halstead_difficulty;
        self.halstead_level += other.halstead_level;
        self.halstead_effort += other.halstead_effort;
        self.halstead_time += other.halstead_time;
        self.halstead_bugs += other.halstead_bugs;
        self.sloc += other.sloc;
        self.ploc += other.ploc;
        self.lloc += other.lloc;
        self.cloc += other.cloc;
        self.blank += other.blank;
        self.functions += other.functions;
        self.closures += other.closures;
        self.mi_original += other.mi_original;
        self.mi_sei += other.mi_sei;
        self.mi_visual_studio += other.mi_visual_studio;
    }
}

File: analyzer/src/metrics.py
from __future__ import annotations
from analyzer.src.values import Values
from typing import Any, Dict, List, Optional
import json
from enum import Enum


class Metric(Enum):
    """Enum containing the path to each value in the data dictionary."""
    NARGS = ("nargs", "sum")
    NEXITS = ("nexits", "sum")
    COGNITIVE = ("cognitive", "sum")
    CYCLOMATIC = ("cyclomatic", "sum")
    U_OPERATORS = ("halstead", "n1")
    OPERATORS = ("halstead", "N1")
    U_OPERANDS = ("halstead", "n2")
    OPERANDS = ("halstead", "N2")
    LENGTH = ("halstead", "length")
    ESTIMATED_PROGRAM_LENGTH = ("halstead", "estimated_program_length")
    PURITY_RATIO = ("halstead", "purity_ratio")
    VOCABULARY = ("halstead", "vocabulary")
    VOLUME = ("halstead", "volume")
    DIFFICULTY = ("halstead", "difficulty")
    LEVEL = ("halstead", "level")
    EFFORT = ("halstead", "effort")
    TIME = ("halstead", "time")
    BUGS = ("halstead", "bugs")
    FUNCTIONS = ("nom", "functions")
    CLOSURES = ("nom", "closures")
    SLOC = ("loc", "sloc")
    PLOC = ("loc", "ploc")
    LLOC = ("loc", "lloc")
    CLOC = ("loc", "cloc")
    BLANK = ("loc", "blank")
    MI_ORIGINAL = ("mi", "mi_original")
    MI_SEI = ("mi", "mi_sei")
    MI_VISUAL_STUDIO = ("mi", "mi_visual_studio")

    @staticmethod
    def as_list() -> List[str]:
        """
        Returns a list of all metrics.

        :return: List of all metrics
        """
        return list(map(lambda x: x.name.lower(), Metric))

    @staticmethod
    def usability_metrics() -> List[str]:
        """
        Returns a list of metrics which are linked to usability.

        :return: Usability metrics
        """
        usability_metrics = [Metric.COGNITIVE, Metric.CYCLOMATIC, Metric.LENGTH, Metric.ESTIMATED_PROGRAM_LENGTH, Metric.PURITY_RATIO, Metric.VOCABULARY,
                             Metric.VOLUME, Metric.DIFFICULTY, Metric.LEVEL, Metric.EFFORT, Metric.TIME, Metric.BUGS, Metric.MI_ORIGINAL, Metric.MI_SEI, Metric.MI_VISUAL_STUDIO]
        return list(map(lambda x: x.name.lower(), usability_metrics))

    @staticmethod
    def as_dict() -> Dict[str, List[str]]:
        """
        Returns a dict representation of all metrics.

        :return: Dict mapping the metrics to a list of tokens
        """
        return dict(map(lambda x: (x.name.lower(), list(x.value)), Metric))


class Metrics:
    """This class represents the whole metric suite and offers a range of utility methods."""

    def __init__(self, data: Optional[Dict[str, Any]] = None):
        for name, path in Metric.as_dict().items():
            if data:
                values = Values([data[path[0]][path[1]]])
            else:
                values = Values([])
            setattr(self, name, values)

    def __str__(self) -> str:
        """
        Prints the metrics.

        :return: Pretty printed JSON string
        """
        return json.dumps(self.as_dict(), indent=4)

    def __repr__(self) -> str:
        """Returns a string representation.

        :return: Pretty printed JSON string
        """
        return json.dumps(self.as_dict(), indent=4)

    def as_dict(self) -> Dict[str, Any]:
        """
        Returns a dict representation.

        :return: Dict containing all metrics
        """
        return {k: getattr(self, k).as_dict() for k in Metric.as_list()}

    def merge(self, other: Metrics) -> None:
        """
        Merges two metric suites.

        :param other: The other metrics
        """
        for name in Metric.as_list():
            self_metric: Values = getattr(self, name)
            other_metric: Values = getattr(other, name)

            self_metric.merge(other_metric)
